### PR TITLE
feat(#892): search and filter stage templates

### LINKED ISSUE #892: Search and filter stage templates
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind feature

**What happened**:

**What you expected to happen**:

Stage templates can be searched by name and filtered by type.

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:
/assign
/priority p0

### PR DESCRIPTION
**Pull request by supereagle**
**What this PR does / why we need it**:

Support to search and filter stage templates

**Which issue(s) this PR is related to** *(optional, link to 3rd issue(s))*:

Fixes #892

Reference to #

**Special notes for your reviewer**:

/cc @zhujian7  @cd1989 

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/server/handler/v1alpha1/template.go [hunk 1]
```diff
@@ -45,7 +54,59 @@ func ListTemplates(ctx context.Context, tenant string, includePublic bool, query
...
+		// Support multiple filters rules: name or alias, and type, separated with comma.
+		filterParts := strings.Split(query.Filter, ",")
+		filters := make(map[string]string)
+		for _, part := range filterParts {
+			kv := strings.Split(part, "=")
+			if len(kv) != 2 {
+				return nil, cerr.ErrorQueryParamNotCorrect.Error(query.Filter)
+			}
+
+			filters[kv[0]] = strings.ToLower(kv[1])
```
Comment (cd1989) [inline pkg/server/handler/v1alpha1/template.go]: Why convert to lowercase, we can't filter case sensitively if so ?

File: pkg/server/handler/v1alpha1/template.go [hunk 2]
```diff
@@ -45,7 +54,59 @@ func ListTemplates(ctx context.Context, tenant string, includePublic bool, query
...
+			selected = true
+			for key, value := range filters {
+				switch key {
+				case "name":
+					if !strings.Contains(item.Name, value) {
+						selected = false
+					}
+				case "alias":
+					if item.Annotations != nil {
+						if alias, ok := item.Annotations[meta.AnnotationAlias]; ok {
```
Comment (cd1989) [inline pkg/server/handler/v1alpha1/template.go]: If user specified `alias` filter, but a item doesn't have annotation alias, will it be accepted ? I think here if `item.Annotations` is nil or `ok` is false, also need set `selected` to false.

File: pkg/server/handler/v1alpha1/template.go [hunk 3]
```diff
@@ -45,7 +54,59 @@ func ListTemplates(ctx context.Context, tenant string, includePublic bool, query
...
+							if !strings.Contains(alias, value) {
+								selected = false
+							}
+						}
+					}
+				case "type":
+					if item.Labels != nil {
```
Comment (cd1989) [inline pkg/server/handler/v1alpha1/template.go]: The same above. If no correct label found, should set `selected` to false.

File: pkg/server/handler/v1alpha1/template.go [hunk 4]
```diff
@@ -45,7 +54,59 @@ func ListTemplates(ctx context.Context, tenant string, includePublic bool, query
 		items = append(items, publicTemplates.Items...)
 	}
 
-	size := int64(len(items))
+	results := []v1alpha1.Stage{}
+	if query.Filter == "" {
+		results = items
+	} else {
```
Comment (cd1989) [inline pkg/server/handler/v1alpha1/template.go]: Suggest to put `else` part to a new function.

File: pkg/server/handler/v1alpha1/template.go (hunk 1 above)
Comment (supereagle) [reply pkg/server/handler/v1alpha1/template.go]: It is hard to user to make sure the case in filter, so we only provide case insensitive filter.

**Comment by supereagle** [comment]
Have fixed comments, @cd1989  PTAL

**Pull request merged**